pg_search/src/postgres/types.rs

```diff
@@ -83,19 +83,20 @@ impl TantivyValue {
         }
     }
 
-    fn json_value_to_tantivy_value(value: Value) -> Vec<TantivyValue> {
-        let mut tantivy_values = vec![];
-        match value {
-            // A tantivy JSON value can't be a top-level array, so we have to make
-            // separate values out of each entry.
-            Value::Array(value_vec) => {
-                for value in value_vec {
-                    tantivy_values.extend_from_slice(&Self::json_value_to_tantivy_value(value));
-                }
-            }
-            _ => tantivy_values.push(TantivyValue(tantivy::schema::OwnedValue::from(value))),
+    fn json_value_to_tantivy_value(value: Value) -> Box<dyn Iterator<Item = TantivyValue>> {
```

Comment:
One cool Rust trick to get around boxing different variants of a `dyn Trait`, if you know their concrete forms, is to instead use an enum.  It ends up being quite a bit faster because it gets rid of the dynamic dispatch behind `dyn`.

What you could do here, roughly, is:

```rust
enum JsonValueIter {
   Many(std::vec::IntoIter<std::iter::FlatMap<TantivyValue>>),
   One(std::iter::Once<TantivyValue>)
}

impl Iterator for JsonValueIter {
    type Item = TantivyValue;

    fn next(&mut self) -> Option<Self::Item> {
       match self {
          Self::Many(iter) => iter.next(),
          Self::One(iter) => iter.next(),
       }
    }
    
    // similar for size_hint() and count() too -- generally good to also overload those
}
```

Then this function would return `-> JsonValueIter` and you'd just return:

```rust
   if ... {
      JsonValueIter::Many(value_vec.into_iter().flat_map(...))
  } else {
      JsonValueIter::Once(std::iter::once(...))
  }
```

I've used this pattern *a lot* and it both reduces memory consumption b/c no more boxing and it's almost always measurably faster under the profiler b/c dynamic dispatch is gone.